Feature/enhance backup script

OpenSpaghettiSauce opened the pull request:
Sorry that I make another appearance regarding this script. I use it a lot and the latest THREADS was a great improvement.

This time, I added two things:

- NODATE
This skips the creation of the timestamp folder. Why? If you back up via s3fs, or backblaze/s3 client, you may opt for overwriting your backup files, and store them with the same name. This way, you can use bucket policy to practically never run out of disk space (well, never run out of money) and still have several snapshots available.

Usage: Specifying any value, such as NODATE=1 will work.
- 'help'
Adds a new option the user can call to see what the script can actually do.
It can be called with ".sh help", or if the user simply calls the script without any argument, the script now tells them about this.

**Feel free to change NODATE**'s name, or the text of help, of course. This worked for me, but I'm just a user of this great project. Thank you guys for your continued work on Mailcow!

OpenSpaghettiSauce committed:
Add support for NODATE env

This adds a THREADS like env var and it makes the script skip the creation of the ${DATE} folder.

OpenSpaghettiSauce committed:
Introduce 'help'

This adds a "help" text, so the user knows their options. Calling just the script will now tell the user to either user 'backup', 'restore', or 'help'.

DerLinkman commented:
Nice nice. But one question i have left somewhere in my brain :D : 

+ What happen if you restore this backup with the NO_DATE Variable? How does the Script behave then?

OpenSpaghettiSauce commented:
Okay that is an absolutely valid question! I'll test it out ASAP and report back/fix my branch.

OpenSpaghettiSauce commented:
Totally valid point.

I suppose the var support could be added here too but...
But, if I think about it more... why even check for a folder?

I mean, shouldn't the user just point the script to a folder, such as let's say `/backup/mailcow-2022-11-03-01-49-04/`?

I am saying this with utmost respect, this is your work, your project.
Just asking, like maybe it'd easier to drop the whole folder selection thing?

For sanity check, like "to make sure this is a proper backup folder", I would probably change the script again (yeah, I know, I should just close my account, sorry). Let's say put a "mailcow_backup.txt" within the folder, And the content would be like:
`for var in *; do md5sum $var; done`. And if the files in the folder check out, then it is a valid backup?

```
md5val backup_crypt.tar.gz
md5val backup_mariadb.tar.gz
md5val backup_postfix.tar.gz
md5val backup_redis.tar.gz
md5val backup_rspamd.tar.gz
md5val backup_vmail.tar.gz
md5val mailcow.conf
```

But then you also have the specific backup modes, like when you only backup some part, and not 'all'.

I am just thinking out loud, so to say. Let me know if you have any ideas...

DerLinkman commented:
> I suppose the var support could be added here too but...

Yeah i think this is the way we want to go.

Not all users want to use a bucket for backup savings and therefore the policys.

So i guess to make everyone happy we´ll simply add the variable.

OpenSpaghettiSauce commented:
I'll add it then in a moment. Thank you.
Well I'm saying a moment but I swear I'll get 10 calls in the next 5 minutes. They sniff this out.

DerLinkman commented:
> I'll add it then in a moment. Thank you. Well I'm saying a moment but I swear I'll get 10 calls in the next 5 minutes. They sniff this out.

Good man. I actually was about trying to do this as well. But please feel free.

OpenSpaghettiSauce commented:
No u. 

Seriously, it's your project, you know how you like this to be implemented. And thank you for not forgetting about this, appreciate it!

DerLinkman commented:
> No u. 
> 
> 
> 
> Seriously, it's your project, you know how you like this to be implemented. And thank you for not forgetting about this, appreciate it! 

Yes but please do that though I've deleted the things (which were not many to be honest) because I'm not sure how to implement it.

OpenSpaghettiSauce committed:
Fix restore for NODATE

I tested it in Virtualbox with my own live server backups and it seemed to restore fine like this. I'm not sure if the help/explanation is enough, feel free to add more, change wording, etc.

DerLinkman commented on helper-scripts/backup_and_restore.sh:
```diff
@@ -367,10 +371,20 @@ if [[ ${1} == "backup" ]]; then
+      exit 1
+    fi
+  else
+    if [[ $(find ${BACKUP_LOCATION}/* -maxdepth 1 -type d 2> /dev/null| wc -l) -lt 1 ]]; then
```
Hm.

It seems not to work on my side:

```
/opt/mailcow-dockerized# NODATE=1 THREADS=8 ./helper-scripts/backup_and_restore.sh restore
Backup location (absolute path, starting with /): /opt/backups
Using 8 Thread(s) for this run.
Notice: You can set the Thread count with the THREADS Variable before you run this script.
Using /opt/backups as backup/restore location.

Found project name mailcowdockerized
Selected backup location has no subfolders
```

The location looks like that:
```
/opt/mailcow-dockerized# NODATE=1 THREADS=8 ./helper-scripts/backup_and_restore.sh restore
Backup location (absolute path, starting with /): /opt/backups
Using 8 Thread(s) for this run.
Notice: You can set the Thread count with the THREADS Variable before you run this script.
Using /opt/backups as backup/restore location.

Found project name mailcowdockerized
Selected backup location has no subfolders
```

OpenSpaghettiSauce replied:
I will work on this ASAP, sorry. I swear I tested it in Virtualbox and it restored 100% fine. I bet it's the end `/` character or something, will verify/dig.

DerLinkman reviewed (changes requested):
See comment :)

andryyy commented:
Besides that it's "mailcow" and it needs a doc PR.

OpenSpaghettiSauce commented:
Sorry @andryyy , could you elaborate, please? Thank you!

OpenSpaghettiSauce committed:
Fixed restore with NODATE

Apparently I changed the problematic for loop on the VM, but I forgot to copy the script back to the git folder before committing. I put the for loop into the IF branches now.

TESTING IS UNDERWAY - I will comment again once it is 100% verified working (for me, again, sorry.)

OpenSpaghettiSauce commented:
Okay so I tested it, it worked for me (NODATE=1), please verify once time allows. Thank you.

DerLinkman commented:
> Sorry @andryyy , could you elaborate, please? Thank you!

What @andryyy was saying is that firstly the "Mailcow" Names inside your PR have to be called "mailcow" with a small M and secondly that this new Parameter needs a entry in the mailcow docs repo: https://github.com/mailcow/mailcow-dockerized-docs

OpenSpaghettiSauce committed:
Fix project name spelling

mailcow is mailcow.

milkmaker commented:
This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs.

milkmaker commented:
This pull request has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs.

DaGuich commented:
Why is this PR blocked? Is the only thing missing the documentation?

Programie commented:
Will this PR ever be merged? I'm really interested in the `NODATE` option so I can use the mailcow backup folder in my borg backup without having timestamped paths in my already timestamped borg backups.
